devito/interpolators.py

```diff
@@ -0,0 +1,410 @@
+import sympy
+import numpy as np
+
+from abc import ABC, abstractmethod
+from cached_property import cached_property
+
+from devito.logger import warning
+from devito.symbolics import retrieve_function_carriers, indexify, INT
+from devito.tools import Evaluable, powerset, flatten, prod
+from devito.types import Eq, Inc
+from devito.types.basic import Scalar
+from devito.types.dense import SubFunction
+from devito.types.dimension import ConditionalDimension, Dimension, DefaultDimension
+
+
+class UnevaluatedSparseOperation(Evaluable):
+
+    """
+    Represents an Injection or an Interpolation operation performed on a
+    SparseFunction. Evaluates to a list of Eq objects.
+
+    Parameters
+    ----------
+    interpolator : Interpolator
+        Interpolator object that will be used to evaluate the Operation.
+    *args, **kwargs
+        The arguments passed to the corresponding method.
+    """
+    subdomain = None
+
+    def __init__(self, interpolator, *args, **kwargs):
+        assert(isinstance(interpolator, GenericInterpolator))
+
+        self.interpolator = interpolator
```

Comment:
This is cyclic, ` self.interpolator` is a`Injection/Interpolation` that is an `UnevaluatedSparseOperation`, that doesn't feel right

Reply:
`self.interpolator` is of type `GenericInterpolator`. I specifically assert that in the line above.